### ISSUE TITLE
Lei Mao's Log Book  Making Reinforcement Learning Agent Library

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6185]
Lei Mao's Log Book &ndash; Making Reinforcement Learning Agent Library<br>
<br>
https://ift.tt/37Vf3lU<br>
<br>
Twitter<br>
<br>
<div>
<div>
<h3>Ideas</h3>
<p>When I was developing learning agents using different reinforcement learning algorithms, somehow I made these learning agents into class. All these learning agents take the current environment state, the action for the current environment state, the reward got after taking the action, the next environment state, and also possible the next action for the next environment state, as input. They all output actions when given the environment state. So all these learning agents could be employed and tested in different learning enviroments, especially the OpenAI Gym environment.</p>
<p><br></p>
<p>I have already implemented several learning agents and prepared the corresponding enviroment running script in my <a href="https://github.com/leimao/OpenAI">GitHub</a>. One could simply change the enviroment name and the learning agent to import in the environment running script to test.</p>
<p><br></p>
<p>I just came up with this idea. So there is still no documentation of how to use these learning agents on your own computer. The Python expert might find easy to use it because you only have to change a little bit in the raw code in order to make it work. However, the Python beginners might have to wait me to get some time to arrange these learning agents to more formal classes and write official documentations.</p>
<p><br></p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/rl.jpg"></p>
<h3>Talk About Algorithms</h3>
<h4>Actor-Critic Policy Gradient</h4>
<p>In the <a href="https://leimao.github.io/article/REINFORCE-Policy-Gradient.html">REINFORCE Policy Gradient reinforcement learning algorithm</a>, we calculate the value of the next state after taking the action $G_t$ ($v_t$) from the rewards in the whole episode and use it to guide our gradient descent. So it is a statistically sampled value. One could actually construct another neural network to represent the value of the state and update it routinely until we get the a good estimate of the true value of the state.</p>
<p><br></p>
<p>The policy network output the action and use the value of the next state to guide its future actions. So the policy network is just like an actor learning how to perform. The value network use the reward after the actor taking the action to estimate the value of the state, and provide this information to the actor. So the value network is just like an critic informing whether the actor did good or bad. This circus is called &ldquo;Actor-Critic&rdquo; Policy Gradient method.</p>
<p><br></p>
<p>The <a href="http://incompleteideas.net/sutton/book/the-book-2nd.html">Sutton Book draft</a> provided the pseudocode for one-step Actor-Critic method. However it is episodic (see below).</p>
<p><br></p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/actor-critic_episodic.png"></p>
<p><br></p>
<p>I did not find a pseudocode continuing case in this draft. However, we just have to modify this pseudocode a little bit to get the pseudocode for the continuing case. To do this, simply remove everything that is related to I (see below).</p>
<p><br></p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/actor-critic_continuing.jpg"></p>
<p><br></p>
<p>I should also mention that the value we used to guide our gradient descent is different to what we used in REINFORCE. This value is called time-dependent error (td_error). It will not affect the direction of our gradient descent mathematically but tune the step width to make our training have less variance (in principal). Please see &ldquo;REINFORCE with Baseline&rdquo; in the Sutton Book draft for more details.</p>
<p><br></p>
<p>Despite the update for the critic value network, everything is very similar to the REINFORCE Policy Gradient method.</p>
<p><br></p>
<p>I tested the algorithm in the CartPole environment and uploaded to my GitHub:</p>
<p><a href="https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Actor-Critic/2017-05-08-v1">https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Actor-Critic/2017-05-08-v1</a></p>
<p><br></p>
<p>The learning performance of the algorithm was submitted to OpenAI Gym:</p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/performance_ac.png"></p>
<p><a href="https://gym.openai.com/evaluations/eval_T51WU12pRlWydRLLI9bdg">https://gym.openai.com/evaluations/eval_T51WU12pRlWydRLLI9bdg</a></p>
<h4>Sarsa Actor-Critic Policy Gradient</h4>
<p>Sarsa Actor-Critic Policy Gradient is a variant of the Actor-Critic Policy Gradient I talked above. It uses Q-network instead of the value network. You may remember the Q-network is a network to represent the value of state-action pairs. To update the Q-network, instead of using DQN that I used for <a href="https://leimao.github.io/journal/Flappy-Bird-AI.html">Flappy Bird</a>, people tend to use Sarsa (?), a variant method to update Q-network. I should have introduced Sarsa before I introduce Sarsa Actor-Critic Policy Gradient. But when I was doing the implementation, I implemented Sarsa from the Sarsa Actor-Critic Policy Gradient template. I got used to present things in a time-dependent manner, because it is time-saving and easy to tell a story.</p>
<p><br></p>
<p>The <a href="http://www0.cs.ucl.ac.uk/staff/D.Silver/web/Teaching.html">Silver Courseware</a> provided the pseudocode for one-step Sarsa Actor-Critic method. It is weird that he called it QAC though it does Sarsa.</p>
<p><br></p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/Sarsa_AC.png"></p>
<p><br></p>
<p>It should also be noted that there is slight gradient descent detail that is different to the Actor-Critic Policy Gradient method I mentioned above. It might be tricky to understand, but it is correct. I have not thought of a straightforward way to explain this to you. So I just skipped it for now.</p>
<p><br></p>
<p>I tested the algorithm in the CartPole environment and uploaded to my GitHub:</p>
<p><a href="https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Sarsa_Actor-Critic/2017-05-09-v1">https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Sarsa_Actor-Critic/2017-05-09-v1</a></p>
<p><br></p>
<p>The learning performance of the algorithm was submitted to OpenAI Gym:</p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/performance_sarsa_ac.png"></p>
<p><a href="https://gym.openai.com/evaluations/eval_fiRpY1pESl2KJCq9nsRq3w">https://gym.openai.com/evaluations/eval_fiRpY1pESl2KJCq9nsRq3w</a></p>
<h4>Sarsa</h4>
<p>The basic concept of Sarsa is almost the same to the critic value network I mentioned above, except for the sampling of action. Sarsa Actor-Critic Policy Gradient uses softmax to sample action, Sarsa here use $\epsilon$-greedy to sample action.</p>
<p><br></p>
<p>The <a href="http://incompleteideas.net/sutton/book/the-book-2nd.html">Sutton Book draft</a> provided the pseudocode for one-step episodic Sarsa method (see below).</p>
<p><br></p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/Sarsa_episodic.png"></p>
<p><br></p>
<p>I tested the algorithm in the CartPole environment and uploaded to my GitHub:</p>
<p><a href="https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Sarsa/2017-05-09-v1">https://github.com/leimao/OpenAI/tree/master/OpenAI_Gym_Solutions/CartPole-v0/Sarsa/2017-05-09-v1</a></p>
<p><br></p>
<p>The learning performance of the algorithm was submitted to OpenAI Gym:</p>
<p><img src="https://leimao.github.io/images/article/2017-05-09-Reinforcement-Learning-Agent-Library/performance_sarsa.png"></p>
<p><a href="https://gym.openai.com/evaluations/eval_dgUQcF9tSiioTjqMyVOiA">https://gym.openai.com/evaluations/eval_dgUQcF9tSiioTjqMyVOiA</a></p>
<p>All the three learning algorithms above are one-stepped. Although they have all solved the problem as is shown in OpenAI Gym evaluations, the still have very big training performance fluctuations, which makes me frustrated. One of my friend told me that I could save the model that has the best averaged training performance during the training, and use it for test. I did that and it worked well. The code is reflected in the code for Sarsa algorithm. To further solve the fluctuation issue, I may try n-stepped version of these algorithms in the future.</p>
<hr>
<footer>
<div>
<img alt="Lei Mao bio photo" src="https://leimao.github.io/images//author_images/Lei-Bio-Medium.jpg">
<h3>Lei Mao</h3>
<p>Machine Learning, Artificial Intelligence, Computer Science.</p>
<a href="http://twitter.com/matchaleimao" target="_blank">Twitter</a> <a href="http://facebook.com/dukeleimao" target="_blank"> Facebook</a> <a href="http://linkedin.com/in/lei-mao" target="_blank"> LinkedIn</a> <a href="http://github.com/leimao" target="_blank"> GitHub</a> <a href="http://scholar.google.com/citations?user=R2VUf7YAAAAJ" target="_blank">&nbsp; G. Scholar</a> <a href="mailto:dukeleimao@gmail.com" target="_blank"> E-Mail</a> <a href="https://leimao.github.io/feed.xml" target="_blank">RSS</a>
</div>
<p><strong>Making Reinforcement Learning Agent Library</strong> was published on <time datetime="2017-05-09T00:00:00-05:00">May 09, 2017</time> and last modified on <time datetime="2017-05-09">May 09, 2017</time> by <a title="About Lei Mao" href="https://leimao.github.io">Lei Mao</a>.</p>
</footer>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Lei Mao's Log Book https://ift.tt/2MtcO1I<br>
<br>
January 13, 2020 at 05:53PM
